File: ts/src/detect.ts
/**
 * ROAR Protocol — Protocol auto-detection adapter (Layer 4).
 *
 * Examines the structure of an incoming JSON message to determine whether
 * it is ROAR native, MCP (JSON-RPC 2.0), A2A (Google A2A), or ACP protocol
 * format, then provides a best-effort normalisation to ROARMessage.
 *
 * Detection priority:
 *   1. ROAR   — has "roar" version field and "intent" field
 *   2. A2A    — has "role" field whose value is "user" | "assistant" | "system"
 *   3. ACP    — JSON-RPC 2.0 with method starting with "session/" | "initialize" | "shutdown"
 *   4. MCP    — JSON-RPC 2.0 with "method" field (generic)
 *   5. UNKNOWN
 *
 * Note: ACP is checked before generic MCP because ACP is a JSON-RPC 2.0 subset.
 */

import { randomHex } from "./types.js";
import type { ROARMessage, AgentIdentity } from "./types.js";
import { MessageIntent } from "./types.js";

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export type DetectedProtocol = "roar" | "mcp" | "a2a" | "acp" | "unknown";

// ---------------------------------------------------------------------------
// Internal constants
// ---------------------------------------------------------------------------

const VALID_INTENTS = new Set<string>(Object.values(MessageIntent));

const A2A_ROLES = new Set(["user", "assistant", "system"]);

const ACP_METHOD_PREFIXES = ["session/", "initialize", "shutdown"];

// ---------------------------------------------------------------------------
// detectProtocol
// ---------------------------------------------------------------------------

/**
 * Detect the protocol of an incoming raw message object.
 *
 * @param raw - The parsed JSON message as a plain object.
 * @returns The detected protocol identifier.
 */
export function detectProtocol(raw: Record<string, unknown>): DetectedProtocol {
  // 1. ROAR: has "roar" version field and valid "intent" field
  if ("roar" in raw && typeof raw["intent"] === "string" && VALID_INTENTS.has(raw["intent"])) {
    return "roar";
  }

  // 2. A2A: has "role" field with a Google A2A role value
  if (typeof raw["role"] === "string" && A2A_ROLES.has(raw["role"])) {
    return "a2a";
  }

  // 3. JSON-RPC 2.0 based protocols
  if (raw["jsonrpc"] === "2.0" && typeof raw["method"] === "string") {
    const method = raw["method"] as string;

    // ACP must be checked before generic MCP (ACP is a JSON-RPC 2.0 subset)
    if (ACP_METHOD_PREFIXES.some((prefix) => method.startsWith(prefix))) {
      return "acp";
    }

    // Generic MCP
    return "mcp";
  }

  return "unknown";
}

// ---------------------------------------------------------------------------
// normalizeToROAR
// ---------------------------------------------------------------------------

/** Build a minimal anonymous AgentIdentity placeholder. */
function anonymousIdentity(displayName: string): AgentIdentity {
  return {
    did: `did:roar:agent:${displayName}-${randomHex(8)}`,
    display_name: displayName,
    agent_type: "agent",
    capabilities: [],
    version: "1.0",
    public_key: null,
  };
}

/**
 * Convert a detected foreign-protocol message to a ROARMessage on a
 * best-effort basis. Returns null for ROAR messages (already native)
 * and for unknown messages that cannot be mapped.
 *
 * @param raw - The parsed JSON message as a plain object.
 * @returns A ROARMessage, or null if the message cannot be normalised.
 */
export function normalizeToROAR(raw: Record<string, unknown>): ROARMessage | null {
  const protocol = detectProtocol(raw);

  switch (protocol) {
    case "roar":
      // Already ROAR — callers should use messageFromWire() directly.
      return null;

    case "mcp": {
      // MCP tool-call request: { jsonrpc, id, method, params }
      const toolName = (raw["method"] as string) ?? "unknown";
      const params = (raw["params"] as Record<string, unknown>) ?? {};
      return {
        roar: "1.0",
        id: `msg_${randomHex(5)}`,
        from_identity: anonymousIdentity("mcp-caller"),
        to_identity: anonymousIdentity(toolName),
        intent: MessageIntent.EXECUTE,
        payload: { action: toolName, params },
        context: { protocol: "mcp", jsonrpc_id: raw["id"] ?? null },
        auth: {},
        timestamp: Date.now() / 1000,
      };
    }

    case "a2a": {
      // A2A chat message: { role, content, ... }
      const role = (raw["role"] as string) ?? "user";
      const content = raw["content"] ?? "";
      const intent = role === "user" ? MessageIntent.ASK : MessageIntent.RESPOND;
      return {
        roar: "1.0",
        id: `msg_${randomHex(5)}`,
        from_identity: anonymousIdentity(`a2a-${role}`),
        to_identity: anonymousIdentity("a2a-agent"),
        intent,
        payload: { content },
        context: { protocol: "a2a", role },
        auth: {},
        timestamp: Date.now() / 1000,
      };
    }

    case "acp": {
      // ACP JSON-RPC 2.0 method call: session/ or initialize / shutdown
      const method = (raw["method"] as string) ?? "";
      const params = (raw["params"] as Record<string, unknown>) ?? {};
      const isSessionEnd =
        method === "shutdown" || method === "session/end" || method === "session/delete";
      const intent = isSessionEnd ? MessageIntent.NOTIFY : MessageIntent.ASK;
      return {
        roar: "1.0",
        id: `msg_${randomHex(5)}`,
        from_identity: anonymousIdentity("acp-ide"),
        to_identity: anonymousIdentity("acp-agent"),
        intent,
        payload: { method, params },
        context: { protocol: "acp", jsonrpc_id: raw["id"] ?? null },
        auth: {},
        timestamp: Date.now() / 1000,
      };
    }

    default:
      return null;
  }
}

File: python/src/roar_sdk/adapters/detect.py
# -*- coding: utf-8 -*-
"""Protocol auto-detection — sniff incoming messages to identify their format.

Examines the structure of an incoming JSON message to determine whether
it's ROAR native, MCP (JSON-RPC 2.0), A2A, or ACP protocol format, then
routes to the appropriate adapter.

Usage::

    from roar_sdk.adapters.detect import detect_protocol, ProtocolType

    msg = json.loads(raw_body)
    protocol = detect_protocol(msg)

    if protocol == ProtocolType.ROAR:
        roar_msg = ROARMessage.model_validate(msg)
    elif protocol == ProtocolType.MCP:
        roar_msg = MCPAdapter.mcp_to_roar(...)
    elif protocol == ProtocolType.A2A:
        roar_msg = A2AAdapter.a2a_task_to_roar(...)
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Dict


class ProtocolType(str, Enum):
    """Detected protocol type."""

    ROAR = "roar"
    MCP = "mcp"
    A2A = "a2a"
    ACP = "acp"
    AUTOGEN = "autogen"
    CREWAI = "crewai"
    LANGGRAPH = "langgraph"
    UNKNOWN = "unknown"


_MCP_METHOD_PREFIXES = (
    "tools/",
    "resources/",
    "prompts/",
    "completion/",
    "initialize",
    "notifications/",
)

_A2A_METHOD_PREFIXES = ("tasks/", "agent/")


def detect_protocol(message: Dict[str, Any]) -> ProtocolType:
    """Detect the protocol of an incoming message.

    Detection heuristics (in priority order):
    1. ROAR: Has "roar" version field and "intent" field
    2. ACP: Has "role" field and "content" field (ACP message body)
    3. MCP: JSON-RPC 2.0 with MCP method prefix
    4. A2A: JSON-RPC 2.0 with tasks/ or agent/ method prefix, or task envelope

    Args:
        message: The raw JSON message dict.

    Returns:
        The detected ProtocolType.
    """
    # ROAR native
    if "roar" in message and "intent" in message:
        return ProtocolType.ROAR

    # AutoGen: role+content with tool_calls, function_call, or function/tool role
    if "role" in message and "content" in message:
        if "tool_calls" in message or "function_call" in message:
            return ProtocolType.AUTOGEN
        if message.get("role") in ("function", "tool") and "name" in message:
            return ProtocolType.AUTOGEN

    # ACP message (session-based, role/content structure — after AutoGen check)
    if "role" in message and "content" in message:
        return ProtocolType.ACP

    # JSON-RPC based protocols
    if message.get("jsonrpc") == "2.0":
        method = message.get("method", "")

        if any(method.startswith(p) for p in _A2A_METHOD_PREFIXES):
            return ProtocolType.A2A

        if any(method.startswith(p) for p in _MCP_METHOD_PREFIXES):
            return ProtocolType.MCP

        # Infer from result structure
        result = message.get("result", {})
        if isinstance(result, dict):
            if "status" in result and "id" in result:
                return ProtocolType.A2A
            if "tools" in result or "resources" in result:
                return ProtocolType.MCP

    # A2A task envelope (no jsonrpc wrapper)
    if "status" in message and "id" in message and "artifacts" in message:
        return ProtocolType.A2A

    # CrewAI: task with "description" + "expected_output"
    if "description" in message and "expected_output" in message:
        return ProtocolType.CREWAI

    # LangGraph: state with "messages" list + "next" node
    if "messages" in message and isinstance(message.get("messages"), list):
        if "next" in message:
            return ProtocolType.LANGGRAPH

    return ProtocolType.UNKNOWN
